### PR TITLE
fix(linter/eslint/no-useless-return): handle switch case continuation

### LINKED ISSUE #23978: linter: no-useless-return false positive on early return after statement in switch case
### What version of Oxlint are you using?

1.71.0

### What command did you run?

oxlint -c 

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```jsonc
{
  "rules": {
    "no-useless-return": "error"
  }
}
```

### What happened?

Oxlint reports no-useless-return on early return statements that are not useless — they prevent subsequent code in the same switch case from running.

[Steps to reproduce](https://playground.oxc.rs/?lintRules=eslint%2Fno-useless-return&options=%7B%22run%22%3A%7B%22lint%22%3Atrue%2C%22formatter%22%3Afalse%2C%22transform%22%3Afalse%2C%22isolatedDeclarations%22%3Afalse%2C%22whitespace%22%3Afalse%2C%22mangle%22%3Afalse%2C%22compress%22%3Afalse%2C%22scope%22%3Atrue%2C%22symbol%22%3Atrue%2C%22cfg%22%3Afalse%7D%2C%22parser%22%3A%7B%22extension%22%3A%22tsx%22%2C%22allowReturnOutsideFunction%22%3Atrue%2C%22preserveParens%22%3Atrue%2C%22allowV8Intrinsics%22%3Atrue%2C%22semanticErrors%22%3Atrue%7D%2C%22linter%22%3A%7B%22config%22%3A%22%7B%5C%22categories%5C%22%3A%7B%5C%22correctness%5C%22%3A%5C%22off%5C%22%7D%2C%5C%22rules%5C%22%3A%7B%5C%22eslint%2Fno-useless-return%5C%22%3A%5C%22error%5C%22%7D%7D%22%7D%2C%22formatter%22%3A%7B%22useTabs%22%3Afalse%2C%22tabWidth%22%3A2%2C%22endOfLine%22%3A%22lf%22%2C%22printWidth%22%3A80%2C%22singleQuote%22%3Afalse%2C%22jsxSingleQuote%22%3Afalse%2C%22quoteProps%22%3A%22as-needed%22%2C%22trailingComma%22%3A%22all%22%2C%22semi%22%3Atrue%2C%22arrowParens%22%3A%22always%22%2C%22bracketSpacing%22%3Atrue%2C%22bracketSameLine%22%3Afalse%2C%22objectWrap%22%3A%22preserve%22%2C%22singleAttributePerLine%22%3Afalse%7D%2C%22transformer%22%3A%7B%22target%22%3A%22es2015%22%2C%22useDefineForClassFields%22%3Atrue%2C%22experimentalDecorators%22%3Atrue%2C%22emitDecoratorMetadata%22%3Atrue%2C%22optimizeEnums%22%3Atrue%2C%22optimizeConstEnums%22%3Atrue%7D%2C%22isolatedDeclarations%22%3A%7B%22stripInternal%22%3Afalse%7D%2C%22codegen%22%3A%7B%22normal%22%3Atrue%2C%22jsdoc%22%3Atrue%2C%22annotation%22%3Atrue%2C%22legal%22%3Atrue%7D%2C%22compress%22%3A%7B%7D%2C%22mangle%22%3A%7B%22topLevel%22%3Atrue%2C%22keepNames%22%3Afalse%7D%2C%22controlFlow%22%3A%7B%22verbose%22%3Afalse%7D%2C%22inject%22%3A%7B%22inject%22%3A%7B%7D%7D%2C%22define%22%3A%7B%22define%22%3A%7B%7D%7D%7D&code=import+React%2C+%7B+useEffect%2C+useRef+%7D+from+%27react%27%0A%0A%0Aconst+changeStep+%3D+%28%29+%3D%3E+%7B%0A++switch+%28direction%29+%7B%0A++++case+DIRECTION.BACKWARD%3A%0A++++++if+%28step+%3D%3D%3D+STEPS.Step1%29+%7B%0A++++++++setIsFlowShown%28false%29%0A++++++++return%0A++++++%7D%0A++++++setStep%281%29%0A++++++break%0A++++case+DIRECTION.FORWARD%3A%0A++++++if+%28step+%3D%3D%3D+STEPS.Step5%29+%7B%0A++++++++setIsFlowShown%28false%29%0A++++++++return%0A++++++%7D%0A++++++setStep%282%29%0A++++++break%0A++%7D%0A%7D%0A%0Aexport+default+DummyComponent)

### PR DESCRIPTION
**Pull request by camc314**
## Summary
- use the CFG unreachable continuation after a bare `return;` to detect code made reachable by removing it
- keep terminal continuations like `break;` and bare `return;` from making the original return useful
- add a regression case for switch branches that continue with additional statements

Fixes #23978

### CODE REVIEW

File: crates/oxc_linter/src/rules/eslint/no_useless_return.rs [hunk 1]
```diff
@@ -273,6 +282,74 @@ impl NoUselessReturn {
...
+            }));
+        }
+
+        false
+    }
+
+    fn block_continuation_state(block: &BasicBlock) -> ContinuationState {
+        for instruction in block.instructions() {
+            match instruction.kind {
+                InstructionKind::Statement
```
[the remark this answers is not in the document]
Comment (camc314) [reply crates/oxc_linter/src/rules/eslint/no_useless_return.rs]: Fixed in b22d24e092: the continuation walk now checks the `InstructionKind::Statement` node id and skips `AstKind::EmptyStatement`, so `; break` remains a useless-return case. I also added the exact regression and updated the snapshot.

File: crates/oxc_linter/src/rules/eslint/no_useless_return.rs [hunk 2]
```diff
@@ -273,6 +286,112 @@ impl NoUselessReturn {
...
+                    | EdgeType::Unreachable => Some(ContinuationBlock {
+                        id: edge.target(),
+                        in_finalizer: block.in_finalizer,
+                    }),
+                    EdgeType::Finalize => {
+                        Some(ContinuationBlock { id: edge.target(), in_finalizer: true })
+                    }
+                    EdgeType::Join => {
+                        Some(ContinuationBlock { id: edge.target(), in_finalizer: false })
+                    }
```
[the remark this answers is not in the document]
Comment (camc314) [reply crates/oxc_linter/src/rules/eslint/no_useless_return.rs]: Fixed in 1802cd3278: the continuation walk now tracks finalizer depth instead of a boolean, so an inner `Join` only exits the nested finalizer and preserves the outer finalizer state. Added the nested `try/finally` regression and updated the snapshot.

File: crates/oxc_linter/src/rules/eslint/no_useless_return.rs [hunk 3]
```diff
@@ -273,6 +286,119 @@ impl NoUselessReturn {
...
+        in_finalizer: bool,
+        nodes: &AstNodes,
+    ) -> ContinuationState {
+        for instruction in block.instructions() {
+            match instruction.kind {
+                InstructionKind::Statement
+                | InstructionKind::Condition
+                | InstructionKind::Iteration(_) => {
+                    if !in_finalizer
+                        && !Self::is_empty_statement_instruction(instruction.node_id, nodes)
```
[the remark this answers is not in the document]
Comment (camc314) [reply crates/oxc_linter/src/rules/eslint/no_useless_return.rs]: Fixed in 6bdc5dd5cb: the continuation helper now treats `AstKind::BlockStatement` as no-op when its body is entirely no-op statements, so `{}` followed by terminal `break` no longer suppresses the `no-useless-return` diagnostic. Added the exact switch regression and updated the snapshot.

File: crates/oxc_linter/src/rules/eslint/no_useless_return.rs [hunk 4]
```diff
@@ -273,6 +286,131 @@ impl NoUselessReturn {
...
+            // finalizer path and resume normal detection after the `Join` edge.
+            stack.extend(graph.edges_directed(block_id, Direction::Outgoing).filter_map(|edge| {
+                match edge.weight() {
+                    EdgeType::Normal
+                    | EdgeType::Jump
+                    | EdgeType::Backedge
+                    | EdgeType::Unreachable => Some(ContinuationBlock {
+                        id: edge.target(),
+                        finalizer_depth: block.finalizer_depth,
+                    }),
```
[the remark this answers is not in the document]
Comment (camc314) [reply crates/oxc_linter/src/rules/eslint/no_useless_return.rs]: Fixed in f0ed232f53: normal continuation edges now recompute finalizer depth from the target block’s AST finalizer ancestors, so a `try/catch/finally` finalizer can exit via `Normal` and code after it is no longer ignored. Added the switch + try/catch/finally regression as a passing case.

File: crates/oxc_linter/src/rules/eslint/no_useless_return.rs [hunk 5]
```diff
@@ -273,6 +286,178 @@ impl NoUselessReturn {
...
+            }
+        }
+
+        ContinuationState::Passthrough
+    }
+
+    fn is_noop_statement_instruction(node_id: Option<NodeId>, nodes: &AstNodes) -> bool {
+        node_id.is_some_and(|node_id| match nodes.kind(node_id) {
+            AstKind::EmptyStatement(_) => true,
+            AstKind::BlockStatement(block) => block.body.iter().all(Self::is_noop_statement),
```
[the remark this answers is not in the document]
Comment (camc314) [reply crates/oxc_linter/src/rules/eslint/no_useless_return.rs]: Handled by the rewrite in 675228dfb8. The CFG continuation walker was removed; switch continuations are now classified with AST tail effects, and `BlockStatement` delegates to `block_statement_effect`, so `{ break; }` is classified as `Stops` rather than meaningful.

File: crates/oxc_linter/src/rules/eslint/no_useless_return.rs [hunk 6]
```diff
@@ -174,22 +185,23 @@ impl NoUselessReturn {
...
                         let is_last_case = idx == switch_stmt.cases.len() - 1;
 
-                        if !is_last_case
-                            && case.consequent.last().is_some_and(|last_stmt| {
-                                last_stmt.span().contains_inclusive(current_span)
+                        if falls_through
+                            && !is_last_case
+                            && switch_stmt.cases.iter().skip(idx + 1).any(|case| {
+                                Self::case_consequent_has_meaningful_effect(&case.consequent)
                             })
```
[the remark this answers is not in the document]
Comment (camc314) [reply crates/oxc_linter/src/rules/eslint/no_useless_return.rs]: Resolved by the CFG-continuation rewrite in b4d9f2810d plus the targeted switch-case condition fix in 446b078c28. The rule no longer scans later switch cases with AST tail helpers; it walks the CFG unreachable continuation from the deleted return. When that walk enters a switch-case condition block via fallthrough, it now follows only the `Jump` edge into that case body, so `case 2: break; default: work();` stops at the `break` instead of considering `default` reachable. Added this exact regression to the fail cases and updated the snapshot.

File: crates/oxc_linter/src/rules/eslint/no_useless_return.rs [hunk 7]
```diff
@@ -273,6 +284,39 @@ impl NoUselessReturn {
...
     }
+
+    fn is_meaningful_statement(node_id: Option<NodeId>, ctx: &LintContext) -> bool {
+        let Some(node_id) = node_id else {
+            return true;
+        };
+
+        match ctx.nodes().kind(node_id) {
+            AstKind::EmptyStatement(_) | AstKind::Function(_) => false,
+            AstKind::BlockStatement(block) => !block.body.iter().all(Self::is_noop_statement),
```
[the remark this answers is not in the document]
Comment (camc314) [reply crates/oxc_linter/src/rules/eslint/no_useless_return.rs]: Fixed in 83e3594739.  CFG instructions are now treated as structural rather than meaningful; their inner instructions determine the effect, so  reaches the inner  and stops before . Added the exact regression as a failing rule case and updated the snapshot.

File: crates/oxc_linter/src/rules/eslint/no_useless_return.rs [hunk 8]
```diff
@@ -228,38 +206,71 @@ impl NoUselessReturn {
...
+            stack.extend(graph.edges_directed(block_id, Direction::Outgoing).filter_map(|edge| {
+                let is_continuation_edge = if is_switch_case_condition_block {
+                    matches!(edge.weight(), EdgeType::Jump)
+                } else {
+                    matches!(
+                        edge.weight(),
+                        EdgeType::Normal
+                            | EdgeType::Jump
+                            | EdgeType::Backedge
+                            | EdgeType::Unreachable
```
[the remark this answers is not in the document]
Comment (camc314) [reply crates/oxc_linter/src/rules/eslint/no_useless_return.rs:283]: Fixed in 83e3594739. The continuation walk now follows  and  edges, but ignores instructions whose AST ancestry places them inside the finalizer because those execute with or without the original return. Terminal instructions inside the finalizer stop that path; once the walk leaves the finalizer, statements such as  count as newly reachable. Added the exact switch + empty-finally regression as a passing case.

[the remark this answers is not in the document]
Comment (camc314) [reply crates/oxc_linter/src/rules/eslint/no_useless_return.rs:283]: Correction: fixed in 83e3594739. The walk follows EdgeType::Finalize and EdgeType::Join, ignores instructions syntactically inside the finalizer, and then recognizes setStep(1) after the finalizer as newly reachable. Terminal finalizer instructions stop the path. The exact switch plus empty-finally case is covered as a passing regression.

File: crates/oxc_linter/src/rules/eslint/no_useless_return.rs [hunk 9]
```diff
@@ -273,6 +300,53 @@ impl NoUselessReturn {
...
+
+    fn is_switch_case_condition_block(
+        cfg: &ControlFlowGraph,
+        ctx: &LintContext,
+        block_id: BlockNodeId,
+    ) -> bool {
+        cfg.basic_block(block_id).instructions().iter().any(|instr| {
+            matches!(instr.kind, InstructionKind::Condition)
+                && instr.node_id.is_some_and(|node_id| {
+                    matches!(ctx.nodes().parent_kind(node_id), AstKind::SwitchCase(_))
```
[the remark this answers is not in the document]
Comment (camc314) [reply crates/oxc_linter/src/rules/eslint/no_useless_return.rs:346]: Fixed in 60285f9669. Switch continuation entry detection now covers both tested case entries and empty default entries with a Jump edge into their body. Fallthrough into default therefore follows the body edge, reaches the break, and does not scan into later cases. Added the exact default-before-case regression and updated the snapshot.

**Comment by camc314** [comment]
### Merge activity

* **Jul 3, 3:14 PM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Jul 3, 3:14 PM UTC**: `camc314` added this pull request to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).
* **Jul 3, 3:23 PM UTC**: Merged by the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).
